src/thermo/BinarySolutionTabulatedThermo.cpp

```diff
@@ -211,27 +316,66 @@ void BinarySolutionTabulatedThermo::initThermoXML(XML_Node& phaseNode, const std
     ThermoPhase::initThermoXML(phaseNode, id_);
 }
 
-std::pair<double, double> BinarySolutionTabulatedThermo::interpolate(double x) const
+double BinarySolutionTabulatedThermo::interpolate(double x, const vector_fp& molefrac,
+                                                  const vector_fp& inputData) const
 {
-    std::pair<double, double> c;
+    double c;
     // Check if x is out of bound
-    if (x > m_molefrac_tab.back()) {
-        c.first = m_enthalpy_tab.back();
-        c.second = m_entropy_tab.back();
+    if (x > molefrac.back()) {
+        c = inputData.back();
         return c;
     }
-    if (x < m_molefrac_tab[0]) {
-        c.first = m_enthalpy_tab[0];
-        c.second = m_entropy_tab[0];
+    if (x < molefrac.front()) {
+        c = inputData.front();
         return c;
     }
-    size_t i = std::distance(m_molefrac_tab.begin(),
-      std::lower_bound(m_molefrac_tab.begin(), m_molefrac_tab.end(), x));
-    c.first = m_enthalpy_tab[i-1] + (m_enthalpy_tab[i] - m_enthalpy_tab[i-1])
-      * (x - m_molefrac_tab[i-1])/(m_molefrac_tab[i]- m_molefrac_tab[i-1]);
-    c.second = m_entropy_tab[i-1] + (m_entropy_tab[i] - m_entropy_tab[i-1])
-      * (x - m_molefrac_tab[i-1])/(m_molefrac_tab[i]- m_molefrac_tab[i-1]);
+    size_t i = std::distance(molefrac.begin(),
+            std::lower_bound(molefrac.begin(), molefrac.end(), x));
+    c = inputData[i-1] + (inputData[i] - inputData[i-1])
+            * (x - molefrac[i-1]) / (molefrac[i] - molefrac[i-1]);
     return c;
 }
 
+void BinarySolutionTabulatedThermo::derive(vector_fp& inputData, vector_fp& derivedData,
+                                           vector_fp& moleFraction)
+{
+    if (inputData.size() > 1) {
+        derivedData[0] = (inputData[1] - inputData[0]) /
+                (moleFraction[1] - moleFraction[0]);
+        derivedData.back() = (inputData.back() - inputData[inputData.size()-2]) /
+                (moleFraction.back() - moleFraction[moleFraction.size()-2]);
+
+        if (inputData.size() > 2) {
+            for (size_t i = 1; i < inputData.size()-1; i++) {
+                derivedData[i] = (inputData[i+1] - inputData[i-1]) /
+                        (moleFraction[i+1] - moleFraction[i-1]);
+            }
+        }
+    }
+    else {
```

Comment:
A minor formatting suggestion:
```suggestion
    } else {
```